add test asserting correct number of columns in results

PeterDSteinberg opened the pull request:
This PR adds a test to assert we have the correct number of items in each row of each table and that the first item is a string on each row while others are float or int.

Here's an example of failing test for METTR which apparently relates to the [webapp-public issue 337 of NA's](https://github.com/OpenSourcePolicyCenter/webapp-public/issues/337) (the UI doesn't have enough row items to render):
```
_________ test_correct_number_of_cols[industry_mettr-baseline-table19] _________

industry_or_asset = 'industry_mettr', changed_reform_base = 'baseline'
table = [['METTR, corporate typically financed investment', 'METTR, non-corporate typically financed investment', 'METTR, corp...mining', 0.3523797776249587, 0.090973254207698209, 0.42973967056296619, 0.14750994372353893, -0.1195836930034178], ...]

    @pytest.mark.needs_puf
    @pytest.mark.parametrize('industry_or_asset,changed_reform_base,table', pytest_args)
    def test_correct_number_of_cols(industry_or_asset, changed_reform_base, table):
        '''Test that each table row has 7 elements where the first is a string
        and others are floats or ints'''
        lens = set(map(len, table[1:]))
        most_lens = tuple(lens)[0]
>       assert len(lens) == 1 and most_lens == 7 and len(table[0]) == most_lens - 1
E       assert (1 == 1 and 6 == 7)
E        +  where 1 = len(set([6]))
```

But note there are other failures, with some tables having too many columns.  

@jdebacker Can you look into this with:
`py.test btax/tests/test_correct_col_counts.py`

jdebacker commented:
@PeterDSteinberg , I was able to run the test, but I don't know how to read the output to diagnose the issue.  There are some columns that are string variables (e.g. that contain asset or industry names).  But when I look at the csv files output from B-Tax, I don't see any non-numeric data in columns that should be numeric.

PeterDSteinberg commented:
The most critical thing to check is the dictionaries coming from `run_btax_to_json_tables` and that each row of those tables has 7 items where the first is a string and others are float/int.  It's possible that the CSVs are correct and the error is after that somewhere.  For example in the error quoted above, it shows METTR tables are returned with 6 items rather than the 7 items required by the UI.  In other cases the table rows have more than 7.  (7 includes the row label and 6 float columns seen in the UI).

jdebacker commented:
Ok- I do see a few places with missing values for assets (something I'm in the process of tracking down and fixing).  Maybe that's the cause.  I'll look more into these dictionaries.

jdebacker commented:
@PeterDSteinberg  - my commit here fixed 12 of the 18 errors.  They had to do with some data entry mistakes in the json files for the output tables.

jdebacker commented:
@PeterDSteinberg , just one error left.  I still can't decipher the messages to determine where the error is, but I think it may  be due to the fact that the depreciation table will have 9 columns when don't by industry and only 7 when done by asset.  All other tables have 7 columns.  Can you adjust test to account for this and see if we then get a pass?  Thanks.

PeterDSteinberg commented:
Here's what I am getting @jdebacker: 

A number of tables with wrong number of columns:
```
=================================== FAILURES ===================================
_____________ test_correct_number_of_cols[asset_d-changed-table12] _____________

industry_or_asset = 'asset_d', changed_reform_base = 'changed'
table = [['NPV of depreciation deductions, typically financed corporate investment', 'NPV of depreciation deductions, typicall..., -0.0072300180179982743, ...], ['Terminals', -0.0021380245491499084, 0.0, 0.0, 0.0, -0.0072300180179982743, ...], ...]

    @pytest.mark.needs_puf
    @pytest.mark.parametrize('industry_or_asset,changed_reform_base,table', pytest_args)
    def test_correct_number_of_cols(industry_or_asset, changed_reform_base, table):
        '''Test that each table row has 7 elements where the first is a string
        and others are floats or ints'''
        if industry_or_asset == 'industry_d':
            expected_cols = 9
        else:
            expected_cols = 7
        lens = set(map(len, table[1:]))
        most_lens = tuple(lens)[0]
>       assert len(lens) == 1 and most_lens == expected_cols and len(table[0]) == most_lens - 1
E       assert (1 == 1 and 8 == 7)
E        +  where 1 = len(set([8]))

btax/tests/test_correct_col_count.py:27: AssertionError
____________ test_correct_number_of_cols[asset_d-baseline-table13] _____________

industry_or_asset = 'asset_d', changed_reform_base = 'baseline'
table = [['NPV of depreciation deductions, typically financed corporate investment', 'NPV of depreciation deductions, typicall...ls', 0.94767008669505515, 0.95001918435873567, 0.9398682252837105, 0.94443561827692246, 0.96539942058483907, ...], ...]

    @pytest.mark.needs_puf
    @pytest.mark.parametrize('industry_or_asset,changed_reform_base,table', pytest_args)
    def test_correct_number_of_cols(industry_or_asset, changed_reform_base, table):
        '''Test that each table row has 7 elements where the first is a string
        and others are floats or ints'''
        if industry_or_asset == 'industry_d':
            expected_cols = 9
        else:
            expected_cols = 7
        lens = set(map(len, table[1:]))
        most_lens = tuple(lens)[0]
>       assert len(lens) == 1 and most_lens == expected_cols and len(table[0]) == most_lens - 1
E       assert (1 == 1 and 8 == 7)
E        +  where 1 = len(set([8]))

btax/tests/test_correct_col_count.py:27: AssertionError
_____________ test_correct_number_of_cols[asset_d-reform-table14] ______________

industry_or_asset = 'asset_d', changed_reform_base = 'reform'
table = [['NPV of depreciation deductions, typically financed corporate investment', 'NPV of depreciation deductions, typicall...als', 0.94553206214590524, 0.95001918435873567, 0.9398682252837105, 0.94443561827692246, 0.9581694025668408, ...], ...]

    @pytest.mark.needs_puf
    @pytest.mark.parametrize('industry_or_asset,changed_reform_base,table', pytest_args)
    def test_correct_number_of_cols(industry_or_asset, changed_reform_base, table):
        '''Test that each table row has 7 elements where the first is a string
        and others are floats or ints'''
        if industry_or_asset == 'industry_d':
            expected_cols = 9
        else:
            expected_cols = 7
        lens = set(map(len, table[1:]))
        most_lens = tuple(lens)[0]
>       assert len(lens) == 1 and most_lens == expected_cols and len(table[0]) == most_lens - 1
E       assert (1 == 1 and 8 == 7)
E        +  where 1 = len(set([8]))

btax/tests/test_correct_col_count.py:27: AssertionError
____________ test_each_param_has_effect_slow[btax_depr_25yr_exp-v6] ____________

k = 'btax_depr_25yr_exp'
v = {'description': 'Rate of bonus depreciation on 25-year class property', 'long_name': 'Rate of bonus depreciation on 25-year property', 'max': [-100.0], 'min': [100.0], ...}

    @pytest.mark.parametrize('k,v', [(k,v) for k,v in DEFAULTS
                                      if not ('depr' in k and 'Switch' in k)])
    @pytest.mark.needs_puf
    @pytest.mark.slow
    def test_each_param_has_effect_slow(k, v):
>       tst_each_param_has_effect('slow', k, v)

btax/tests/test_params_have_effect.py:78:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
btax/tests/test_params_have_effect.py:70: in tst_each_param_has_effect
    tst_once(fast_or_slow, **user_mods)

```
The test which should be skipped (see also #92 )
```_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

fast_or_slow = 'slow'
user_params = {'btax_depr_25yr_exp': 0.0, 'btax_depr_25yr_tax_Switch': True}
tables = {'asset_coc': {'baseline': [['Cost of capital, typically financed corporate investment', 'Cost of capital, typically f...134194113977897, 0.17363377676833808, 0.32081707699194256, 0.21672089695453647, -0.19063106125483525, ...], ...]}, ...}
has_changed = False, k = 'industry_d'
changed = [['NPV of depreciation deductions, typically financed corporate investment', 'NPV of depreciation deductions, typicall...cept oil and gas', 0.0, 0.0, 0.0, 0.0, 0.0, ...], ['Support activities for mining', 0.0, 0.0, 0.0, 0.0, 0.0, ...], ...]
row = ['Other services, except government', 0.0, 0.0, 0.0, 0.0, 0.0, ...]
item = 0.0, @py_format1 = 'assert False'

    def tst_once(fast_or_slow, **user_params):
        if fast_or_slow == 'slow':
            # actually run the model
            # and look at the "changed" tables
            tables = run_btax_to_json_tables(**user_params)
            has_changed = False
            for k in tables:
                changed = tables[k]['changed']

                for row in changed:
                    for item in row:
                        if isinstance(item, (float, int)) and item:
                            has_changed = True
                            break
                    if has_changed:
                        break
                if has_changed:
                    break
>           assert has_changed
E           assert False

btax/tests/test_params_have_effect.py:26: AssertionError
----------------------------- Captured stdout call -----------------------------
You loaded data for 2009.
Your data have been extrapolated to 2013.
You loaded data for 2009.
Your data have been extrapolated to 2013.
('base_output_by_asset', 'baseline', u'metr')
('base_output_by_asset', 'baseline', u'mettr')
('base_output_by_asset', 'baseline', u'coc')
('base_output_by_asset', 'baseline', u'd')
('reform_output_by_asset', 'reform', u'metr')
('reform_output_by_asset', 'reform', u'mettr')
('reform_output_by_asset', 'reform', u'coc')
('reform_output_by_asset', 'reform', u'd')
('changed_output_by_asset', 'changed', u'metr')
('changed_output_by_asset', 'changed', u'mettr')
('changed_output_by_asset', 'changed', u'coc')
('changed_output_by_asset', 'changed', u'd')
('base_output_by_industry', 'baseline', u'metr')
('base_output_by_industry', 'baseline', u'mettr')
('base_output_by_industry', 'baseline', u'coc')
('base_output_by_industry', 'baseline', u'd')
('reform_output_by_industry', 'reform', u'metr')
('reform_output_by_industry', 'reform', u'mettr')
('reform_output_by_industry', 'reform', u'coc')
('reform_output_by_industry', 'reform', u'd')
('changed_output_by_industry', 'changed', u'metr')
('changed_output_by_industry', 'changed', u'mettr')
('changed_output_by_industry', 'changed', u'coc')
('changed_output_by_industry', 'changed', u'd')
['base_output_by_asset']
==================== 4 failed, 59 passed in 587.43 seconds =====================
(webapp)
psteinberg at 0178-psteinberg in ~/Documents/AEI/ps_btax on assert-col-count [?]
$
```

I'll go ahead and skip that test.

PeterDSteinberg commented:
Latest commit fixes #92

jdebacker commented:
@peterdsteinberg I'm on the road so just looking on my phone, but this looks right. Errors w depreciation tables who iv have more columns than other tables (I think by asset has one more and by industry two more).

talumbau commented:
@PeterDSteinberg @jdebacker when I update the puf91 test data to be as in PR #93 and run the tests I get the following error:

```
__________________________________________________ ERROR collecting btax/tests/test_correct_col_count.py __________________________________________________
btax/tests/test_correct_col_count.py:11: in <module>
    btax_betr_pass=.3)
btax/run_btax.py:142: in run_btax_to_json_tables
    quit()
../../miniconda3/envs/devpy27/lib/python2.7/site.py:369: in __call__
    raise SystemExit(code)
E   SystemExit: None
--------------------------------------------------------------------- Captured stdout ---------------------------------------------------------------------
You loaded data for 2009.
Your data have been extrapolated to 2013.
You loaded data for 2009.
Your data have been extrapolated to 2013.
('base_output_by_asset', 'baseline', u'metr')
('base_output_by_asset', 'baseline', u'mettr')
('base_output_by_asset', 'baseline', u'coc')
('base_output_by_asset', 'baseline', u'd')
('reform_output_by_asset', 'reform', u'metr')
('reform_output_by_asset', 'reform', u'mettr')
('reform_output_by_asset', 'reform', u'coc')
('reform_output_by_asset', 'reform', u'd')
('changed_output_by_asset', 'changed', u'metr')
('changed_output_by_asset', 'changed', u'mettr')
('changed_output_by_asset', 'changed', u'coc')
('changed_output_by_asset', 'changed', u'd')
('base_output_by_industry', 'baseline', u'metr')
('base_output_by_industry', 'baseline', u'mettr')
('base_output_by_industry', 'baseline', u'coc')
('base_output_by_industry', 'baseline', u'd')
('reform_output_by_industry', 'reform', u'metr')
('reform_output_by_industry', 'reform', u'mettr')
('reform_output_by_industry', 'reform', u'coc')
('reform_output_by_industry', 'reform', u'd')
('changed_output_by_industry', 'changed', u'metr')
('changed_output_by_industry', 'changed', u'mettr')
('changed_output_by_industry', 'changed', u'coc')
('changed_output_by_industry', 'changed', u'd')
['base_output_by_asset']
================================================ 19 tests deselected by "-m 'not slow and not needs_puf'" =================================================
=================================================== 20 passed, 19 deselected, 1 error in 58.07 seconds ====================================================
```

so I still think this PR has an issue by inlining the call to `run_btax` in the collection phase. I don't think that will work because of the explicit exit in that function.

PeterDSteinberg commented:
@jdebacker Does the /ccc UI expect those other columns?  My understanding is that /ccc was developed with the idea all tables have the same number of columns and the column names are hard-coded.  I am not sure what the UI would do with the 8th and 9th columns, if given.  cc @brendancol

PeterDSteinberg commented on btax/run_btax.py:
```diff
@@ -140,7 +140,6 @@ def run_btax_to_json_tables(test_run=False,start_year=2016,iit_reform=None,**use
         else:
             raise ValueError('Expected an "asset" or "industry" related table')
     print['base_output_by_asset']
-    quit()
```
@talumbau Are you sure you pulled the most recent code?  The latest commits remove this `quit`.  You are getting SystemExit error from this quit.  Not sure where it came from.

jdebacker commented:
@TJ, @PeterDSteinberg  Passing Travis CI, but getting 3 failed tests when running `py.test tests/test_correct_col_count.py` locally.   Again, having trouble diagnosing, but:

```

industry_or_asset = 'industry_d', changed_reform_base = 'reform'
table = [['NPV of depreciation deductions, typically financed corporate investment', 'NPV of depreciation deductions, typicall...0019594379109208549, 0.81304868846680756, 0.0019353763903269107, 0.80301377477898339, 0.0020165689280599755, ...], ...]

    @pytest.mark.needs_puf
    @pytest.mark.parametrize('industry_or_asset,changed_reform_base,table', pytest_args)
    def test_correct_number_of_cols(industry_or_asset, changed_reform_base, table):
        '''Test that each table row has 7 elements where the first is a string
        and others are floats or ints'''
        if industry_or_asset == 'industry_d':
            expected_cols = 9
        else:
            expected_cols = 7
        lens = set(map(len, table[1:]))
        most_lens = tuple(lens)[0]
>       assert len(lens) == 1 and most_lens == expected_cols and len(table[0]) == most_lens - 1
E       assert (1 == 1 and 7 == 9)
E        +  where 1 = len(set([7]))

tests/test_correct_col_count.py:27: AssertionError
==================== 3 failed, 21 passed in 8.14 seconds ====================
```

Suggests related to depreciation tables?  Tried removing `delta` columns from depreciation tables to make same number of columns and still not change test results (not including these changes in PR since no effect on test).

PeterDSteinberg commented:
@jdebacker The 3 test failures show that several of the tables are being created with not the expected number of columns.  

With `industry_or_asset = 'asset_d', changed_reform_base = 'changed'`, the expectation is that there will be 7 columns but there are actually 8.  Same pattern with`industry_or_asset = 'asset_d'` and with `changed_reform_base = 'baseline'` and `industry_or_asset = 'asset_d', changed_reform_base = 'reform'`.  We have an extra column in the `asset_d` and `industry_d` tables

jdebacker commented:
@PeterDSteinberg @talumbau  - I think we are good now.  modified test and tables so all have seven columns and column count and Travis CI both pass.  Let me know if ok and I'll merge.
